### PR TITLE
Add support to authenticate with application credential

### PR DESCRIPTION
**Pull request by zioproto**
For #1214 

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

- https://github.com/openstack/keystoneauth/blob/stable/rocky/keystoneauth1/identity/v3/application_credential.py#L48-L67
 - https://github.com/openstack/keystone/blob/stable/rocky/keystone/conf/constants.py#L20
 - https://github.com/openstack/python-keystoneclient/blob/stable/rocky/keystoneclient/v3/application_credentials.py#L42

### CODE REVIEW

**Comment by jtopjian** [comment]
@zioproto Travis is flagging a `go fmt` error. If you can, run [`go fmt`](https://blog.golang.org/go-fmt-your-code) on the changed files - that should fix things up.

**Comment by zioproto** [comment]
@rdl87  I pushed a new version of the patch that works with any possible combination of ID or NAME env vars. Can you review it ? Are you working on the tests ?

**Comment by RdL87** [comment]
@zioproto i'm working on this just now. So i'm going to recompile the client with the new library you've just pushed. I will ping you as soon as the keystone become available.

**Comment by zioproto** [comment]
@jtopjian I believe we are ready for review. I added two tests. I tried to to give correct code pointers to keystone codebase. thank you

File: auth_options.go [hunk 1]
```diff
@@ -194,8 +209,35 @@ func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[s
 			req.Auth.Identity.Token = &tokenReq{
 				ID: opts.TokenID,
 			}
+
+		} else if opts.ApplicationCredentialID != "" {
+			// Configure the request for ApplicationCredentialID authentication.
+			// https://github.com/openstack/keystone-specs/blob/master/specs/keystone/queens/application-credentials.rst#using-an-application-credential-to-obtain-a-token
```
Comment (jtopjian) [inline auth_options.go]: Let's replace the above URL with:

https://github.com/openstack/keystoneauth/blob/ebe781a3ea0386d6ff088a84e8dde26e538b856d/keystoneauth1/identity/v3/application_credential.py#L48-L67

File: auth_options.go [hunk 2]
```diff
@@ -194,8 +209,35 @@ func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[s
...
+			// There are three kinds of possible application_credential requests
+			// 1. application_credential id + secret.
+			// 2. application_credential name + user_id + secret
+			// 3. application_credential name + user_name + user_domain_id/user_domain_name
```
Comment (jtopjian) [inline auth_options.go]: How about rewording this to:

1. application_credential id + secret
2. application_credential name + secret + user_id
3. application_credential name + secret + username + domain_id / domain_name

Secondly, since an app credential secret is always required (I _think_ it's required, right? https://github.com/openstack/keystoneauth/blob/c2ae9e298e2acdc4e653bce18735294cf4df2454/keystoneauth1/loading/_plugins/identity/v3.py#L270-L272), right below this line, let's make sure one is set:

```go
if opts.ApplicationCredentialSecret == "" {
  return nil, ErrAppCredMissingSecret{}
}
```

You can add the new error to the bottom of this file: https://github.com/gophercloud/gophercloud/blob/master/errors.go#L453

File: auth_options.go [hunk 3]
```diff
@@ -194,8 +209,35 @@ func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[s
...
+			req.Auth.Identity.Methods = []string{"application_credential"}
+			req.Auth.Identity.ApplicationCredential = &applicationCredentialReq{
+				ID:     &opts.ApplicationCredentialID,
+				Secret: &opts.ApplicationCredentialSecret,
+			}
+		} else if opts.ApplicationCredentialName != "" {
```
Comment (jtopjian) [inline auth_options.go]: Again, right below this line, let's make sure a secret was provided - same way as above.

In addition, let's also make sure that only one of `DomainName` or `DomainID` were provided, the same way it's done in other places in this file:

```go
if opts.DomainID == "" && opts.DomainName == "" {
  return nil, ErrDomainIDOrDomainName{}
}
```

File: auth_options.go [hunk 4]
```diff
@@ -194,8 +209,35 @@ func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[s
...
+			req.Auth.Identity.Methods = []string{"application_credential"}
+			if opts.DomainID != "" {
+				userRequest = userReq{Name: &opts.Username, Domain: &domainReq{ID: &opts.DomainID}}
```
Comment (jtopjian) [inline auth_options.go]: nit: use multiple lines:

```go
userRequest = userReq{
  Name: &opts.Username,
  ...
}
```

Same for the other occurrences.

File: auth_options.go [hunk 5]
```diff
@@ -194,8 +209,35 @@ func (opts *AuthOptions) ToTokenV3CreateMap(scope map[string]interface{}) (map[s
...
+			} else if opts.DomainName != "" {
+				userRequest = userReq{Name: &opts.Username, Domain: &domainReq{Name: &opts.DomainName}}
+			} else {
+				userRequest = userReq{Name: &opts.Username}
```
Comment (jtopjian) [inline auth_options.go]: I don't think this variation is possible?

Comment (zioproto) [reply auth_options.go]: Probably you are right. @RdL87 I will drop this

Comment (zioproto) [reply auth_options.go]: Yes you are right this cannot work without domain. If we had implemented using UsedID that it would have make sense to create a request without DomainID/Name, but it is not our case.

**Comment by zioproto** [comment]
Thanks for the review. I pushed changes that should address all the points of the review

**Comment by zioproto** [comment]
@jtopjian I am not sure what the problem with the coveralls test not passing

**Comment by jtopjian** [comment]
@zioproto You can ignore Coveralls. It acts weird sometimes.
